### ISSUE TITLE
AUTO_DUPLICATE_LVL doesn't work

### ISSUE THREAD

**Issue by drTr0jan** [OCSInventory-NG/OCSInventory-Server#49]
### General informations
Operating system : FreeBSD 10.3-RELEASE-p11

### Server informations
Perl version : v5.20.3
Mysql  version :  5.7.15

### OCS Inventory informations
Ocs server version : 2.3.0.0 or 2.3.1 (fresh install).

### Problem's description
I have a fresh install of OCS 2.3.0.0, and I've added an _usbdevices_ table to DB.
But AUTO_DUPLICATE_LVL (hostname+uuid) function doesn't work.

Ex. I've one host in OCS with hostname _KBP-WIN8-VM_ and uuid _AA434E54-2E8F-44A4-8C3E-8F9B59AF2B60_ on Windows 8.1 with Agent 2.3.0.0. I've gotten **new** host with same hostname and uuid in OCS after full reinstall of an agent on the host.
I've gotten one error `ERROR *** AGENT => Failed to send Inventory <HTTP Status Code #403>` only after the first run agent on the host. Next runs are without errors.

**UPDATE**
AUTO_DUPLICATE_LVL function doesn't work (without a 403 error) if enabled TRACE_DELETED option.

**Comment by gillesdubois** [comment]
Hi @drTr0jan ,

Your duplicate problem is related to this issue https://github.com/OCSInventory-NG/OCSInventory-Server/issues/50 and will be fixed asap.

Regards,
Gilles Dubois.

**Comment by drTr0jan** [comment]
Please reopen the issue. The problem is not related to the issue #50.

**Comment by phillip-white-sociomantic** [comment]
For me this seems to be fixed in `v2.3.1`. And i have the `usbdevices` table mentioned in https://github.com/OCSInventory-NG/OCSInventory-Server/issues/50

And there are no errors in `/var/log/ocsinventory-server/activity.log` that show any merging errors.

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic is not a #50 bug.
I have a v2.3.1 server and a _usbdevice_ table which empty. And there no errors _/var/log/ocsinventory-server/activity.log_
But AUTO_DUPLICATE_LVL doesn't work and a client has `Failed to send Inventory <HTTP Status Code #403>` error on first run.

**Comment by phillip-white-sociomantic** [comment]
@drTr0jan Which version of the agent / client do you use? I don't have this in the error logs for a Unix `2.2.0` agent. (thats whats int he 16.04 repo)

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic I have an agent v.2.3.0.0 for Windows.

**Comment by phillip-white-sociomantic** [comment]
@drTr0jan strange, i just re-installed `v.2.3.0.0` and don't have this error:
```
==============================================================================
Starting OCS Inventory NG Agent on Friday, March 24, 2017 13:47:06.
AGENT => Running OCS Inventory NG Agent Version 2.3.0.0
AGENT => Using OCS Inventory NG FrameWork Version 2.3.0.0
AGENT => Loading plug-in(s)
AGENT => Using network connection with Communication Server
AGENT => Using Communication Provider <OCS Inventory NG cURL Communication Provider> Version <2.3.0.0>
AGENT => Sending Prolog
AGENT => Prolog successfully sent
SUPPORT => No support detected, Registration key : N/A
AGENT => Inventory required
AGENT => Launching hardware and software checks
```

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic had you delete `C:\ProgramData\OCS Inventory NG\Agent` directory before re-install?

What version of mysql-server do you have?

**Comment by phillip-white-sociomantic** [comment]
@drTr0jan I am running

Server: 2.3.1
Mysql: 5.7.17-11-log Percona Server (GPL), Release '11', Revision 'f60191c'
Windows Agnet: 2.3.0.0

I deleted the `C:\ProgramData\OCS Inventory NG` now, and re-ran the installer. Still it succeeds:

```
==============================================================================
Starting OCS Inventory NG Agent on Monday, March 27, 2017 09:06:54.
AGENT => Running OCS Inventory NG Agent Version 2.3.0.0
AGENT => Using OCS Inventory NG FrameWork Version 2.3.0.0
AGENT => Loading plug-in(s)
AGENT => Using network connection with Communication Server
AGENT => Using Communication Provider <OCS Inventory NG cURL Communication Provider> Version <2.3.0.0>
AGENT => Sending Prolog
DID_CHECK => Hostname changed new:<hostname> old:<>
DID_CHECK => Generating new unique ID for device <hostname>
AGENT => Prolog successfully sent
SUPPORT => No support detected, Registration key : N/A
AGENT => Inventory required
AGENT => Launching hardware and software checks
AGENT => Sending Inventory
INVENTORY => Inventory changed since last run
ADMIN INFOS => Couple ( TAG <=> <my tag> ) added to configuration file
ADMIN INFOS => Couple ( fields_3 <=> 1 ) added to configuration file
ADMIN INFOS => Couple ( fields_5 <=> 1 ) added to configuration file
ADMIN INFOS => Couple ( fields_6 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_7 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_8 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_9 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_10 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_11 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_12 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_13 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_14 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_16 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_18 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_19 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_20 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_21 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_22 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_23 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_24 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_25 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_26 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_27 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_28 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_29 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_30 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_31 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_32 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_34 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_35 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_36 <=>  ) added to configuration file
ADMIN INFOS => Couple ( fields_37 <=>  ) added to configuration file
AGENT => Inventory successfully sent
AGENT =>  Communication Server ask for Package Download
ERROR *** DOWNLOAD => Found result code <SUCCESS> for OCS Inventory Agent Setup package but no package ID specified, so remove all packages to avoid running Agent setup in loop !
DOWNLOAD => Download and setup tool successfully started
AGENT => Unloading communication provider
AGENT => Unloading plug-in(s)
AGENT => Execution duration: 00:00:21.
```

And the server merges the computer fine.

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic hmmm...

Server: 2.3.1
MySQL: 5.7.15-log
Windows agent: 2.3.0.0

But get the log after deleting the `C:\ProgramData\OCS Inventory NG`, and rerunning the installer.

```
==============================================================================
Starting OCS Inventory NG Agent on Monday, March 27, 2017 20:16:08.
AGENT => Running OCS Inventory NG Agent Version 2.3.0.0
AGENT => Using OCS Inventory NG FrameWork Version 2.3.0.0
AGENT => Loading plug-in(s)
AGENT => Using network connection with Communication Server
AGENT => Using Communication Provider <OCS Inventory NG cURL Communication Provider> Version <2.3.0.0>
AGENT => Sending Prolog
DID_CHECK => Hostname changed new:<hostname> old:<>
DID_CHECK => Generating new unique ID for device <hostname>
AGENT => Prolog successfully sent
SUPPORT => No support detected, Registration key : N/A
AGENT => Inventory required
AGENT => Launching hardware and software checks
AGENT => Sending Inventory
INVENTORY => Inventory changed since last run
ERROR *** AGENT => Failed to send Inventory <HTTP Status Code #403>
AGENT =>  Communication Server ask for Package Download
AGENT => Unloading communication provider
AGENT => Unloading plug-in(s)
AGENT => Execution duration: 00:00:35.
```

**Comment by phillip-white-sociomantic** [comment]
@drTr0jan yep, thats exactly what i did. Purge old logs and config and re-install. Strange how you get 403.

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic What logs and config should I purge - server or agent? What should I reinstall?

**Comment by phillip-white-sociomantic** [comment]
@drTr0jan I just removed the `C:\ProgramData\OCS Inventory NG` folder and re-installed the agent. Before i moved the server from `v.2.3.0` to v.2.3.1` i used to see in the error logs:
```
[error] Cannot establish communication : 500 Can't connect to <server> (
Bad hostname '<server>'
```
And then it would still continue to inventory, despite to 500 error.

Since `v2.3.1` i don't have any communication or duplicate errors.

How did you do the upgrade to `v2.3.1`? Maybe something went wrong with it.

**Comment by drTr0jan** [comment]
@phillip-white-sociomantic there wasn't interesting info in the error log (httpd-error) before upgrade from `v.2.3.0` to `v.2.3.1`.

I have upgraded to `v.2.3.1` as follows:
1) Run setup.sh script and answered questions. There were not an errors in ocs_server_setup.log.
2) Restarted the Apache webserver.
2) Gone to the reports page and upgraded databases in browser.

I can try a fresh install of `v.2.3.1` server.

**Comment by drTr0jan** [comment]
I have installed a fresh `v.2.3.1` server but problem has exists.

Also I have inquired about 403 error. It is thrown by next code from `Apache/Ocsinventory/Server/Inventory.pm:97`:
```
  # Lock device
  if(&_lock($Apache::Ocsinventory::CURRENT_CONTEXT{'DATABASE_ID'})){
    &_log( 516, 'inventory', 'device_locked');
    return(APACHE_FORBIDDEN);
  }
```

**Comment by drTr0jan** [comment]
I've found solution!

403 error was caused by next code from `Apache/Ocsinventory/Server/System.pm:212`:
```
  if($dbh->do('INSERT INTO locks(HARDWARE_ID, ID, SINCE) VALUES(?,?,NOW())', {} , $device, $$ )){
    ...
    return(0);
  }else{
    ...
    return 1;
  }
```
`$dbh->do()` had caused an error on the `locks` (MEMORY-type) table : `Statement violates GTID consistency: Updates to non-transactional tables can only be done in either autocommitted statements or single-statement transactions, and never in the same statement as updates to transactional tables.`

I had the environment with enabled `enforce-gtid-consistency` feature in MySQL 5.7. This feature doesn't allow simple update to memory tables.

**Comment by drTr0jan** [comment]
I've disabled `enforce-gtid-consistency`, and the error 403 not more exists. But `AUTO_DUPLICATE_LVL` still doesn't work. 
I've found enabling OCS option `TRACE_DELETED` causes the behavior. If I disable `TRACE_DELETED` option `AUTO_DUPLICATE_LVL` starts work successfully. Why?

**Comment by drTr0jan** [comment]
Last problem (with TRACE_DELETED) is solved in #55.
But the locks problem I suggest to solve in this ticket.

P.S. Need an enhancement tag.
P.P.S. @phillip-white-sociomantic thx for the inspiration. :)

**Comment by gillesdubois** [comment]
Hi, 

I'm checking your locks issue, I'm closing this issue and opening an other one to have a better visibility ( I will also check your auto duplicate fix proposal )

Regards,
Gilles Dubois.
